MRI_geom/spd_fmri.py

```diff
@@ -0,0 +1,185 @@
+        nb_edges = mapping.shape[0]
+        for edge in range(nb_edges):
+            e0, e1 = (mapping.iloc[edge]["mapA"], mapping.iloc[edge]["mapB"])
+            u[map_FCN_r[e0], map_FCN_r[e1]] = graphs.iloc[graph_id][edge]
+            u = np.multiply(u, (u > CORR_THRESH))
+        return np.abs(u + u.T)
+                       
+        
+    for graph_id in range(NB_GRAPHS):
+        all_graphs[graph_id] = create_connectome(graph_id, mapping)
+        all_targets[graph_id] = graph_labels.loc[graphs.index[graph_id], "Class"]
+
+    np.random.seed(RAND_SEED)
+    index_train = range(NB_GRAPHS)
+    np.random.shuffle(index_train)
+    stop = int(0.85 * NB_GRAPHS)
+    labels = np.array(all_targets[index_train])
+    return (graphs.iloc[:, index_train], [all_graphs[t] for t in index_train],
+        labels, stop, index_train)
+
+
+def laplacian(a):
+    d = np.diag(np.array(a.sum(1)).flatten())
+    return d-a
+
+
+def fit_kernel_cv(log_euclidean_distance, labels,
+                  sigma=None, verbose=False):
+    kf = KFold(n_splits=10)
+    perf = {}
+    conf = {}
+    k = 0
+    if sigma is not None:
+        distance = np.exp(- np.square(log_euclidean_distance)/(sigma**2))
+    else:
+        distance = log_euclidean_distance
+    for train_index, test_index in kf.split(range(NB_GRAPHS)):
+        train_index = np.array(train_index)
+        test_index = np.array(test_index)
+        x = np.array(distance[train_index, :])[:, train_index]
+        x_test = np.array(distance[test_index, :])[:, train_index]
+        clf = SVC(kernel='precomputed')
+        clf.fit(x, labels[train_index])
+        y_train = clf.predict(x)
+        if verbose:
+            print("Training accuracy:",
+                  accuracy_score(labels[train_index], y_train))
+        y_test = clf.predict(x_test)
+        perf[k] = {'acc': accuracy_score(labels[test_index], y_test),
+                   'prec': precision_score(labels[test_index], y_test),
+                   'f1': f1_score(labels[test_index], y_test),
+                   'recall': recall_score(labels[test_index], y_test)}
+        conf[k] = confusion_matrix(labels[test_index], y_test)
+        k += 1
+        return perf, conf
+
+
+def main():
+    gamma = GAMMA
+    time_alg = {}
+    graphs, all_graphs, labels, stop, index_train = import_data()
+    distance = {option: np.zeros((NB_GRAPHS, NB_GRAPHS))
+                for option in DISTANCE}
+    tic = time.time()
+    for option in DISTANCE:
+        tic = time.time()
+        for c, g in enumerate(all_graphs):
+            hat_l1 = laplacian(g) + gamma*np.eye(DIM_SPACE)
+            lambda2, u = np.linalg.eigh(hat_l1)
+            inv_l1 = u.dot(np.diag(1.0/np.sqrt(lambda2)).dot(u.T))
+            for j in range(c):
+                hat_l2 = laplacian(all_graphs[j]) + gamma*np.eye(DIM_SPACE)
+                if option == "log_euclidean":
+                    distance[option][c, j] = np.linalg.norm(spd_space.group_log(hat_l1).squeeze(0)
```

Comment:
formatting is really weird here. also 3 for loops is way to much. There has to be a way to do this is a vectorized way